File: folderly.c
#include <dirent.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/stat.h>
#include <unistd.h>

#ifndef PATH_MAX
#define PATH_MAX 4096
#endif

// Limit the number of unique destination folders tracked in memory.
// This keeps the per-run cache small and predictable.
#define MAX_CREATED_FOLDERS 32

typedef struct
{
  const char *ext;
  const char *folder;
} Rule;

// Extension-to-folder map used by the organizer.
// The first matching rule wins, and unmatched files fall back to "others".
static const Rule rules[] = {
    // Documents and text files.
    {".txt", "documents"},
    {".md", "documents"},
    {".markdown", "documents"},
    {".pdf", "documents"},
    {".doc", "documents"},
    {".docx", "documents"},
    {".docm", "documents"},
    {".odt", "documents"},
    {".rtf", "documents"},
    {".tex", "documents"},
    {".epub", "documents"},
    {".mobi", "documents"},
    {".azw", "documents"},
    {".azw3", "documents"},
    {".fb2", "documents"},
    {".djvu", "documents"},
    {".xps", "documents"},
    {".pages", "documents"},
    {".srt", "documents"},
    {".ass", "documents"},
    {".ssa", "documents"},
    {".vtt", "documents"},
    {".sub", "documents"},

    // Spreadsheet files.
    {".csv", "spreadsheets"},
    {".tsv", "spreadsheets"},
    {".xls", "spreadsheets"},
    {".xlsx", "spreadsheets"},
    {".xlsm", "spreadsheets"},
    {".ods", "spreadsheets"},
    {".numbers", "spreadsheets"},

    // Presentation files.
    {".ppt", "presentations"},
    {".pptx", "presentations"},
    {".pptm", "presentations"},
    {".odp", "presentations"},
    {".key", "presentations"},

    // Image and photo files.
    {".jpg", "images"},
    {".jpeg", "images"},
    {".png", "images"},
    {".gif", "images"},
    {".bmp", "images"},
    {".tif", "images"},
    {".tiff", "images"},
    {".webp", "images"},
    {".svg", "images"},
    {".ico", "images"},
    {".heic", "images"},
    {".heif", "images"},
    {".raw", "images"},
    {".cr2", "images"},
    {".nef", "images"},
    {".arw", "images"},
    {".dng", "images"},

    // Audio files.
    {".mp3", "music"},
    {".wav", "music"},
    {".flac", "music"},
    {".aac", "music"},
    {".m4a", "music"},
    {".ogg", "music"},
    {".opus", "music"},
    {".wma", "music"},
    {".aiff", "music"},
    {".alac", "music"},
    {".mid", "music"},
    {".midi", "music"},

    // Video files.
    {".mp4", "videos"},
    {".mkv", "videos"},
    {".avi", "videos"},
    {".mov", "videos"},
    {".wmv", "videos"},
    {".flv", "videos"},
    {".webm", "videos"},
    {".m4v", "videos"},
    {".3gp", "videos"},
    {".mpeg", "videos"},
    {".mpg", "videos"},
    {".ts", "videos"},
    {".m2ts", "videos"},
    {".vob", "videos"},
    {".ogv", "videos"},

    // Archives and disk images.
    {".zip", "archives"},
    {".rar", "archives"},
    {".7z", "archives"},
    {".tar", "archives"},
    {".gz", "archives"},
    {".bz2", "archives"},
    {".xz", "archives"},
    {".tgz", "archives"},
    {".tbz2", "archives"},
    {".txz", "archives"},
    {".zst", "archives"},
    {".cab", "archives"},
    {".lz", "archives"},
    {".lzma", "archives"},
    {".iso", "archives"},

    // Executables and installers.
    {".bat", "programs"},
    {".cmd", "programs"},
    {".com", "programs"},
    {".exe", "programs"},
    {".msi", "programs"},
    {".apk", "programs"},
    {".app", "programs"},
    {".appimage", "programs"},
    {".dmg", "programs"},
    {".pkg", "programs"},
    {".deb", "programs"},
    {".rpm", "programs"},
    {".jar", "programs"},
    {".war", "programs"},
    {".ear", "programs"},
    {".bin", "programs"},
    {".run", "programs"},
    {".so", "programs"},
    {".dll", "programs"},
    {".dylib", "programs"},
    {".sys", "programs"},

    // Source code, build files, scripts, and web files.
    {".c", "code"},
    {".h", "code"},
    {".cpp", "code"},
    {".cc", "code"},
    {".cxx", "code"},
    {".hpp", "code"},
    {".hh", "code"},
    {".java", "code"},
    {".kt", "code"},
    {".kts", "code"},
    {".go", "code"},
    {".rs", "code"},
    {".cs", "code"},
    {".fs", "code"},
    {".swift", "code"},
    {".m", "code"},
    {".mm", "code"},
    {".vb", "code"},
    {".vbs", "code"},
    {".dart", "code"},
    {".scala", "code"},
    {".groovy", "code"},
    {".clj", "code"},
    {".cljs", "code"},
    {".lisp", "code"},
    {".el", "code"},
    {".hs", "code"},
    {".nim", "code"},
    {".zig", "code"},
    {".pas", "code"},
    {".erl", "code"},
    {".ex", "code"},
    {".exs", "code"},
    {".jl", "code"},
    {".r", "code"},
    {".vala", "code"},
    {".cmake", "code"},
    {".gradle", "code"},
    {".make", "code"},
    {".mk", "code"},
    {".sh", "code"},
    {".bash", "code"},
    {".zsh", "code"},
    {".fish", "code"},
    {".py", "code"},
    {".rb", "code"},
    {".php", "code"},
    {".pl", "code"},
    {".pm", "code"},
    {".lua", "code"},
    {".ps1", "code"},
    {".js", "code"},
    {".ts", "code"},
    {".mjs", "code"},
    {".cjs", "code"},
    {".jsx", "code"},
    {".tsx", "code"},
    {".sql", "code"},
    {".html", "code"},
    {".htm", "code"},
    {".xhtml", "code"},
    {".css", "code"},
    {".scss", "code"},
    {".sass", "code"},
    {".less", "code"},

    // Config and structured text files.
    {".json", "configs"},
    {".jsonl", "configs"},
    {".xml", "configs"},
    {".yaml", "configs"},
    {".yml", "configs"},
    {".toml", "configs"},
    {".ini", "configs"},
    {".cfg", "configs"},
    {".conf", "configs"},
    {".properties", "configs"},
    {".env", "configs"},
    {".local", "configs"},
    {".rc", "configs"},
    {".plist", "configs"},
    {".desktop", "configs"},

    // Data and database files.
    {".db", "data"},
    {".db3", "data"},
    {".sqlite", "data"},
    {".sqlite3", "data"},
    {".parquet", "data"},
    {".feather", "data"},
    {".orc", "data"},
    {".h5", "data"},
    {".hdf5", "data"},
    {".npy", "data"},
    {".npz", "data"},
    {".mat", "data"},
    {".pkl", "data"},
    {".pickle", "data"},
    {".sav", "data"},
    {".dat", "data"},

    // Design and creative assets.
    {".psd", "design"},
    {".ai", "design"},
    {".xd", "design"},
    {".fig", "design"},
    {".sketch", "design"},
    {".afdesign", "design"},
    {".indd", "design"},

    // 3D models and scene files.
    {".blend", "3d"},
    {".obj", "3d"},
    {".fbx", "3d"},
    {".stl", "3d"},
    {".dae", "3d"},
    {".glb", "3d"},
    {".gltf", "3d"},
    {".ply", "3d"},
    {".usdz", "3d"},
    {".3ds", "3d"},
    {".amf", "3d"},
    {".3mf", "3d"},

    // Fonts and certificates.
    {".ttf", "fonts"},
    {".otf", "fonts"},
    {".woff", "fonts"},
    {".woff2", "fonts"},
    {".eot", "fonts"},
    {".fon", "fonts"},
    {".pem", "certificates"},
    {".crt", "certificates"},
    {".cer", "certificates"},
    {".key", "certificates"},
    {".pfx", "certificates"},
    {".p12", "certificates"},
    {".der", "certificates"},

    // Logs and backup files.
    {".log", "logs"},
    {".out", "logs"},
    {".err", "logs"},
    {".trace", "logs"},
    {".bak", "backups"},
    {".backup", "backups"},
    {".old", "backups"},
    {".orig", "backups"},
    {".tmp", "backups"},
    {".swp", "backups"},
    {".swo", "backups"},

    {NULL, NULL}};

// Joins two path parts into a single buffer.
// Returns -1 when the output would not fit.
static int build_path(char *buffer, size_t buffer_size, const char *left,
                      const char *right)
{
  if (!buffer || !left || !right || buffer_size == 0)
  {
    return -1;
  }

  int written = snprintf(buffer, buffer_size, "%s/%s", left, right);

  // ! debug log to verify path building and buffer usage
  // printf("---Debug: build_path('%s', '%s') -> '%s' (written: %d)\n", left, right, buffer, written);

  if (written < 0 || (size_t)written >= buffer_size)
  {
    return -1;
  }

  return 0;
}

// Returns the extension, including the dot, or NULL when none exists.
static const char *get_extension(const char *filename)
{
  const char *dot = strrchr(filename, '.');
  if (!dot || dot == filename)
  {
    return NULL;
  }
  return dot;
}

// Maps a file extension to its destination folder.
// Unknown extensions go to the "others" folder.
const char *find_folder(const char *ext)
{
  if (!ext)
  {
    return "others";
  }
  for (int i = 0; rules[i].ext != NULL; i++)
  {
    if (strcasecmp(ext, rules[i].ext) == 0)
    {
      return rules[i].folder;
    }
  }
  return "others";
}

// Creates a directory if it does not already exist.
// Fails when a non-directory item already occupies the path.
static int ensure_dir(const char *path)
{
  struct stat st;
  if (stat(path, &st) == 0)
  {
    return S_ISDIR(st.st_mode) ? 0 : -1;
  }

#ifdef _WIN32
  return mkdir(path);
#else
  return mkdir(path, 0755);
#endif
}

// Moves a file by renaming it into the target folder.
// This is fast because it does not copy file contents.
static int move_file(const char *src, const char *dest)
{
  if (!src || !dest)
  {
    return -1;
  }
  return rename(src, dest);
}

// Removes trailing '/' or '\\' so path joins stay consistent.
static void trim_trailing_separator(char *path)
{
  if (!path)
  {
    return;
  }
  size_t len = strlen(path);
  while (len > 0 && (path[len - 1] == '/' || path[len - 1] == '\\'))
  {
    path[--len] = '\0';
  }
}

// Checks whether the current directory entry is a regular file.
// Uses d_type when available and falls back to stat() otherwise.
static int is_regular_entry(const struct dirent *entry, const char *path)
{
  (void)entry;

#ifdef DT_DIR
  if (entry->d_type == DT_DIR)
  {
    return 0;
  }
#endif

#ifdef DT_REG
  if (entry->d_type == DT_REG)
  {
    return 1;
  }
#endif

  struct stat st;
  return stat(path, &st) == 0 && S_ISREG(st.st_mode);
}

// Creates each destination folder only once per run.
// This avoids repeated stat()/mkdir() calls for common categories.
static int ensure_folder_once(const char *folder, const char *folder_path,
                              const char *created_folders[],
                              size_t *created_count)
{
  for (size_t i = 0; i < *created_count; i++)
  {
    if (strcmp(created_folders[i], folder) == 0)
    {
      return 0;
    }
  }

  if (ensure_dir(folder_path) != 0)
  {
    return -1;
  }

  if (*created_count >= MAX_CREATED_FOLDERS)
  {
    return -1;
  }

  created_folders[*created_count] = folder;
  (*created_count)++;
  return 0;
}

// Scans the target directory and moves each regular file into a category.
// The function keeps subdirectories untouched.
int organize_directory(const char *target_dir)
{
  if (!target_dir)
  {
    fprintf(stderr, "Error: null directory path\n");
    return -1;
  }

  DIR *dir = opendir(target_dir);
  if (!dir)
  {
    perror("Failed to open directory");
    return -1;
  }

  char work_dir[PATH_MAX];
  int copied = snprintf(work_dir, sizeof(work_dir), "%s", target_dir);
  if (copied < 0 || (size_t)copied >= sizeof(work_dir))
  {
    fprintf(stderr, "Error: directory path is too long\n");
    closedir(dir);
    return -1;
  }
  trim_trailing_separator(work_dir);

  struct dirent *entry;
  int moved = 0, failed = 0;
  const char *created_folders[MAX_CREATED_FOLDERS];
  size_t created_count = 0;

  while ((entry = readdir(dir)) != NULL)
  {
    // Skip the current and parent directory entries.
    if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
    {
      continue;
    }

    // Build the full source path for the current entry.
    char source_path[PATH_MAX];
    if (build_path(source_path, sizeof(source_path), work_dir,
                   entry->d_name) != 0)
    {
      failed++;
      continue;
    }

    // Only move regular files, not directories or special entries.
    if (!is_regular_entry(entry, source_path))
    {
      continue;
    }

    // Resolve the folder name from the file extension.
    const char *ext = get_extension(entry->d_name);
    const char *folder = find_folder(ext);

    // Prepare the destination folder path.
    char dest_folder[PATH_MAX];
    if (build_path(dest_folder, sizeof(dest_folder), work_dir, folder) != 0)
    {
      failed++;
      continue;
    }

    if (ensure_folder_once(folder, dest_folder, created_folders,
                           &created_count) != 0)
    {
      fprintf(stderr, "Failed to prepare folder: %s\n", folder);
      failed++;
      continue;
    }

    // Build the final destination path inside the category folder.
    char dest_path[PATH_MAX];
    if (build_path(dest_path, sizeof(dest_path), dest_folder,
                   entry->d_name) != 0)
    {
      failed++;
      continue;
    }

    // rename() is used here because it is fast and avoids file copying.
    if (move_file(source_path, dest_path) == 0)
    {
      printf("Moved: %s -> %s/\n", entry->d_name, folder);
      moved++;
    }
    else
    {
      fprintf(stderr, "Failed to move: %s\n", entry->d_name);
      failed++;
    }
  }

  closedir(dir);
  printf("\nDone! %d moved, %d failed\n", moved, failed);
  return 0;
}

// Program entry point.
// Uses the provided directory path or the current directory by default.
int main(int argc, char *argv[])
{
  printf("Folderly - Simple File Organizer\n- %d", argc);
  const char *target = (argc > 1) ? argv[1] : ".";

  printf("Organizing files in: %s\n\n", target);

  if (organize_directory(target) != 0)
  {
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}

File: folderly.js
#!/usr/bin/env node
'use strict';

const fs = require('fs/promises');
const path = require('path');

// Extension-to-folder map.
// The first matching rule wins, and unmatched files go to "others".
const RULES = [
  // Documents and text files.
  { ext: '.txt', folder: 'documents' },
  { ext: '.md', folder: 'documents' },
  { ext: '.markdown', folder: 'documents' },
  { ext: '.pdf', folder: 'documents' },
  { ext: '.doc', folder: 'documents' },
  { ext: '.docx', folder: 'documents' },
  { ext: '.docm', folder: 'documents' },
  { ext: '.odt', folder: 'documents' },
  { ext: '.rtf', folder: 'documents' },
  { ext: '.tex', folder: 'documents' },
  { ext: '.epub', folder: 'documents' },
  { ext: '.mobi', folder: 'documents' },
  { ext: '.azw', folder: 'documents' },
  { ext: '.azw3', folder: 'documents' },
  { ext: '.fb2', folder: 'documents' },
  { ext: '.djvu', folder: 'documents' },
  { ext: '.xps', folder: 'documents' },
  { ext: '.pages', folder: 'documents' },
  { ext: '.srt', folder: 'documents' },
  { ext: '.ass', folder: 'documents' },
  { ext: '.ssa', folder: 'documents' },
  { ext: '.vtt', folder: 'documents' },
  { ext: '.sub', folder: 'documents' },

  // Spreadsheet files.
  { ext: '.csv', folder: 'spreadsheets' },
  { ext: '.tsv', folder: 'spreadsheets' },
  { ext: '.xls', folder: 'spreadsheets' },
  { ext: '.xlsx', folder: 'spreadsheets' },
  { ext: '.xlsm', folder: 'spreadsheets' },
  { ext: '.ods', folder: 'spreadsheets' },
  { ext: '.numbers', folder: 'spreadsheets' },

  // Presentation files.
  { ext: '.ppt', folder: 'presentations' },
  { ext: '.pptx', folder: 'presentations' },
  { ext: '.pptm', folder: 'presentations' },
  { ext: '.odp', folder: 'presentations' },
  { ext: '.key', folder: 'presentations' },

  // Image and photo files.
  { ext: '.jpg', folder: 'images' },
  { ext: '.jpeg', folder: 'images' },
  { ext: '.png', folder: 'images' },
  { ext: '.gif', folder: 'images' },
  { ext: '.bmp', folder: 'images' },
  { ext: '.tif', folder: 'images' },
  { ext: '.tiff', folder: 'images' },
  { ext: '.webp', folder: 'images' },
  { ext: '.svg', folder: 'images' },
  { ext: '.ico', folder: 'images' },
  { ext: '.heic', folder: 'images' },
  { ext: '.heif', folder: 'images' },
  { ext: '.raw', folder: 'images' },
  { ext: '.cr2', folder: 'images' },
  { ext: '.nef', folder: 'images' },
  { ext: '.arw', folder: 'images' },
  { ext: '.dng', folder: 'images' },

  // Audio files.
  { ext: '.mp3', folder: 'music' },
  { ext: '.wav', folder: 'music' },
  { ext: '.flac', folder: 'music' },
  { ext: '.aac', folder: 'music' },
  { ext: '.m4a', folder: 'music' },
  { ext: '.ogg', folder: 'music' },
  { ext: '.opus', folder: 'music' },
  { ext: '.wma', folder: 'music' },
  { ext: '.aiff', folder: 'music' },
  { ext: '.alac', folder: 'music' },
  { ext: '.mid', folder: 'music' },
  { ext: '.midi', folder: 'music' },

  // Video files.
  { ext: '.mp4', folder: 'videos' },
  { ext: '.mkv', folder: 'videos' },
  { ext: '.avi', folder: 'videos' },
  { ext: '.mov', folder: 'videos' },
  { ext: '.wmv', folder: 'videos' },
  { ext: '.flv', folder: 'videos' },
  { ext: '.webm', folder: 'videos' },
  { ext: '.m4v', folder: 'videos' },
  { ext: '.3gp', folder: 'videos' },
  { ext: '.mpeg', folder: 'videos' },
  { ext: '.mpg', folder: 'videos' },
  { ext: '.m2ts', folder: 'videos' },
  { ext: '.vob', folder: 'videos' },
  { ext: '.ogv', folder: 'videos' },

  // Archives and disk images.
  { ext: '.zip', folder: 'archives' },
  { ext: '.rar', folder: 'archives' },
  { ext: '.7z', folder: 'archives' },
  { ext: '.tar', folder: 'archives' },
  { ext: '.gz', folder: 'archives' },
  { ext: '.bz2', folder: 'archives' },
  { ext: '.xz', folder: 'archives' },
  { ext: '.tgz', folder: 'archives' },
  { ext: '.tbz2', folder: 'archives' },
  { ext: '.txz', folder: 'archives' },
  { ext: '.zst', folder: 'archives' },
  { ext: '.cab', folder: 'archives' },
  { ext: '.lz', folder: 'archives' },
  { ext: '.lzma', folder: 'archives' },
  { ext: '.iso', folder: 'archives' },

  // Executables and installers.
  { ext: '.bat', folder: 'programs' },
  { ext: '.cmd', folder: 'programs' },
  { ext: '.com', folder: 'programs' },
  { ext: '.exe', folder: 'programs' },
  { ext: '.msi', folder: 'programs' },
  { ext: '.apk', folder: 'programs' },
  { ext: '.app', folder: 'programs' },
  { ext: '.appimage', folder: 'programs' },
  { ext: '.dmg', folder: 'programs' },
  { ext: '.pkg', folder: 'programs' },
  { ext: '.deb', folder: 'programs' },
  { ext: '.rpm', folder: 'programs' },
  { ext: '.jar', folder: 'programs' },
  { ext: '.war', folder: 'programs' },
  { ext: '.ear', folder: 'programs' },
  { ext: '.bin', folder: 'programs' },
  { ext: '.run', folder: 'programs' },
  { ext: '.so', folder: 'programs' },
  { ext: '.dll', folder: 'programs' },
  { ext: '.dylib', folder: 'programs' },
  { ext: '.sys', folder: 'programs' },

  // Source code, build files, scripts, and web files.
  { ext: '.c', folder: 'code' },
  { ext: '.h', folder: 'code' },
  { ext: '.cpp', folder: 'code' },
  { ext: '.cc', folder: 'code' },
  { ext: '.cxx', folder: 'code' },
  { ext: '.hpp', folder: 'code' },
  { ext: '.hh', folder: 'code' },
  { ext: '.java', folder: 'code' },
  { ext: '.kt', folder: 'code' },
  { ext: '.kts', folder: 'code' },
  { ext: '.go', folder: 'code' },
  { ext: '.rs', folder: 'code' },
  { ext: '.cs', folder: 'code' },
  { ext: '.fs', folder: 'code' },
  { ext: '.swift', folder: 'code' },
  { ext: '.m', folder: 'code' },
  { ext: '.mm', folder: 'code' },
  { ext: '.vb', folder: 'code' },
  { ext: '.vbs', folder: 'code' },
  { ext: '.dart', folder: 'code' },
  { ext: '.scala', folder: 'code' },
  { ext: '.groovy', folder: 'code' },
  { ext: '.clj', folder: 'code' },
  { ext: '.cljs', folder: 'code' },
  { ext: '.lisp', folder: 'code' },
  { ext: '.el', folder: 'code' },
  { ext: '.hs', folder: 'code' },
  { ext: '.nim', folder: 'code' },
  { ext: '.zig', folder: 'code' },
  { ext: '.pas', folder: 'code' },
  { ext: '.erl', folder: 'code' },
  { ext: '.ex', folder: 'code' },
  { ext: '.exs', folder: 'code' },
  { ext: '.jl', folder: 'code' },
  { ext: '.r', folder: 'code' },
  { ext: '.vala', folder: 'code' },
  { ext: '.cmake', folder: 'code' },
  { ext: '.gradle', folder: 'code' },
  { ext: '.make', folder: 'code' },
  { ext: '.mk', folder: 'code' },
  { ext: '.sh', folder: 'code' },
  { ext: '.bash', folder: 'code' },
  { ext: '.zsh', folder: 'code' },
  { ext: '.fish', folder: 'code' },
  { ext: '.py', folder: 'code' },
  { ext: '.rb', folder: 'code' },
  { ext: '.php', folder: 'code' },
  { ext: '.pl', folder: 'code' },
  { ext: '.pm', folder: 'code' },
  { ext: '.lua', folder: 'code' },
  { ext: '.ps1', folder: 'code' },
  { ext: '.js', folder: 'code' },
  { ext: '.ts', folder: 'code' },
  { ext: '.mjs', folder: 'code' },
  { ext: '.cjs', folder: 'code' },
  { ext: '.jsx', folder: 'code' },
  { ext: '.tsx', folder: 'code' },
  { ext: '.sql', folder: 'code' },
  { ext: '.html', folder: 'code' },
  { ext: '.htm', folder: 'code' },
  { ext: '.xhtml', folder: 'code' },
  { ext: '.css', folder: 'code' },
  { ext: '.scss', folder: 'code' },
  { ext: '.sass', folder: 'code' },
  { ext: '.less', folder: 'code' },

  // Config and structured text files.
  { ext: '.json', folder: 'configs' },
  { ext: '.jsonl', folder: 'configs' },
  { ext: '.xml', folder: 'configs' },
  { ext: '.yaml', folder: 'configs' },
  { ext: '.yml', folder: 'configs' },
  { ext: '.toml', folder: 'configs' },
  { ext: '.ini', folder: 'configs' },
  { ext: '.cfg', folder: 'configs' },
  { ext: '.conf', folder: 'configs' },
  { ext: '.properties', folder: 'configs' },
  { ext: '.env', folder: 'configs' },
  { ext: '.local', folder: 'configs' },
  { ext: '.rc', folder: 'configs' },
  { ext: '.plist', folder: 'configs' },
  { ext: '.desktop', folder: 'configs' },

  // Data and database files.
  { ext: '.db', folder: 'data' },
  { ext: '.db3', folder: 'data' },
  { ext: '.sqlite', folder: 'data' },
  { ext: '.sqlite3', folder: 'data' },
  { ext: '.parquet', folder: 'data' },
  { ext: '.feather', folder: 'data' },
  { ext: '.orc', folder: 'data' },
  { ext: '.h5', folder: 'data' },
  { ext: '.hdf5', folder: 'data' },
  { ext: '.npy', folder: 'data' },
  { ext: '.npz', folder: 'data' },
  { ext: '.mat', folder: 'data' },
  { ext: '.pkl', folder: 'data' },
  { ext: '.pickle', folder: 'data' },
  { ext: '.sav', folder: 'data' },
  { ext: '.dat', folder: 'data' },

  // Design and creative assets.
  { ext: '.psd', folder: 'design' },
  { ext: '.ai', folder: 'design' },
  { ext: '.xd', folder: 'design' },
  { ext: '.fig', folder: 'design' },
  { ext: '.sketch', folder: 'design' },
  { ext: '.afdesign', folder: 'design' },
  { ext: '.indd', folder: 'design' },

  // 3D models and scene files.
  { ext: '.blend', folder: '3d' },
  { ext: '.obj', folder: '3d' },
  { ext: '.fbx', folder: '3d' },
  { ext: '.stl', folder: '3d' },
  { ext: '.dae', folder: '3d' },
  { ext: '.glb', folder: '3d' },
  { ext: '.gltf', folder: '3d' },
  { ext: '.ply', folder: '3d' },
  { ext: '.usdz', folder: '3d' },
  { ext: '.3ds', folder: '3d' },
  { ext: '.amf', folder: '3d' },
  { ext: '.3mf', folder: '3d' },

  // Fonts and certificates.
  { ext: '.ttf', folder: 'fonts' },
  { ext: '.otf', folder: 'fonts' },
  { ext: '.woff', folder: 'fonts' },
  { ext: '.woff2', folder: 'fonts' },
  { ext: '.eot', folder: 'fonts' },
  { ext: '.fon', folder: 'fonts' },
  { ext: '.pem', folder: 'certificates' },
  { ext: '.crt', folder: 'certificates' },
  { ext: '.cer', folder: 'certificates' },
  { ext: '.key', folder: 'certificates' },
  { ext: '.pfx', folder: 'certificates' },
  { ext: '.p12', folder: 'certificates' },
  { ext: '.der', folder: 'certificates' },

  // Logs and backup files.
  { ext: '.log', folder: 'logs' },
  { ext: '.out', folder: 'logs' },
  { ext: '.err', folder: 'logs' },
  { ext: '.trace', folder: 'logs' },
  { ext: '.bak', folder: 'backups' },
  { ext: '.backup', folder: 'backups' },
  { ext: '.old', folder: 'backups' },
  { ext: '.orig', folder: 'backups' },
  { ext: '.tmp', folder: 'backups' },
  { ext: '.swp', folder: 'backups' },
  { ext: '.swo', folder: 'backups' },
];

const RULE_MAP = new Map();
for (const rule of RULES) {
  if (!RULE_MAP.has(rule.ext)) {
    RULE_MAP.set(rule.ext, rule.folder);
  }
}

// Returns the extension, including the dot, or null when none exists.
// Dotfiles like ".env" are treated as extensions when a rule exists.
function getExtension(filename) {
  const baseName = path.basename(filename).toLowerCase();
  if (!baseName || baseName === '.' || baseName === '..') {
    return null;
  }

  const lastDot = baseName.lastIndexOf('.');
  if (lastDot > 0) {
    return baseName.slice(lastDot);
  }

  if (baseName.startsWith('.') && RULE_MAP.has(baseName)) {
    return baseName;
  }

  return null;
}

// Maps a file extension to its destination folder.
// Unknown extensions go to the "others" folder.
function findFolder(ext) {
  if (!ext) {
    return 'others';
  }

  return RULE_MAP.get(ext.toLowerCase()) || 'others';
}

// Removes trailing path separators so joins stay predictable.
function trimTrailingSeparator(input) {
  if (!input) {
    return '.';
  }

  const normalized = path.normalize(input);
  const root = path.parse(normalized).root;
  if (normalized === root) {
    return normalized;
  }

  const trimmed = normalized.replace(/[\\/]+$/, '');
  return trimmed || '.';
}

// Creates a directory if it does not already exist.
async function ensureDir(dirPath, createdFolders) {
  if (createdFolders.has(dirPath)) {
    return;
  }

  await fs.mkdir(dirPath, { recursive: true });
  createdFolders.add(dirPath);
}

// Scans the target directory and moves each regular file into a category.
// Subdirectories are intentionally left untouched.
async function organizeDirectory(targetDir) {
  if (!targetDir) {
    throw new Error('Error: null directory path');
  }

  const workDir = path.resolve(trimTrailingSeparator(targetDir));
  const entries = await fs.readdir(workDir, { withFileTypes: true });
  const createdFolders = new Set();
  let moved = 0;
  let failed = 0;

  for (const entry of entries) {
    if (entry.name === '.' || entry.name === '..') {
      continue;
    }

    const sourcePath = path.join(workDir, entry.name);

    let isFile = entry.isFile();
    if (!isFile) {
      if (entry.isDirectory()) {
        continue;
      }

      try {
        isFile = (await fs.stat(sourcePath)).isFile();
      } catch {
        failed += 1;
        continue;
      }
    }

    if (!isFile) {
      continue;
    }

    let folder = findFolder(getExtension(entry.name));

    // Special rule for .ts files: decide by file size.
    // If file > 1024 KB (1 MB) treat as video, otherwise treat as code.
    const ext = getExtension(entry.name);
    if (ext && ext.toLowerCase() === '.ts') {
      try {
        const st = await fs.stat(sourcePath);
        const bytes = st.size;
        const threshold = 1024 * 1024; // 1024 KB == 1,048,576 bytes
        folder = bytes > threshold ? 'videos' : 'code';
      } catch (err) {
        // if stat fails, fall back to the default folder
      }
    }
    const folderPath = path.join(workDir, folder);

    try {
      await ensureDir(folderPath, createdFolders);
    } catch {
      console.error(`Failed to prepare folder: ${folder}`);
      failed += 1;
      continue;
    }

    const destPath = path.join(folderPath, entry.name);

    try {
      await fs.rename(sourcePath, destPath);
      console.log(`Moved: ${entry.name} -> ${folder}/`);
      moved += 1;
    } catch {
      console.error(`Failed to move: ${entry.name}`);
      failed += 1;
    }
  }

  console.log(`\nDone! ${moved} moved, ${failed} failed`);
  return { moved, failed };
}

async function main(argv = process.argv) {
  const target = argv[2] || '.';

  console.log('Folderly - Simple File Organizer');
  console.log(`Organizing files in: ${target}\n`);

  await organizeDirectory(target);
}

if (require.main === module) {
  main().catch((error) => {
    console.error(error.message || String(error));
    process.exitCode = 1;
  });
}

module.exports = {
  RULES,
  RULE_MAP,
  getExtension,
  findFolder,
  organizeDirectory,
};
